hubmapbags/reports.py

```diff
@@ -0,0 +1,110 @@
+from datetime import datetime
+from pathlib import Path
+
+import pandas as pd
+from pandarallel import pandarallel
+from tqdm import tqdm
+
+from . import apis, utilities
+
+
+def __is_protected(hubmap_id, token=None):
+    metadata = apis.get_dataset_info(hubmap_id, instance="prod", token=token)
+    try:
+        return metadata["contains_human_genetic_sequences"]
+    except:
+        return None
+
+
+def __get_published_timestamp(hubmap_id, token=None):
+    try:
+        metadata = apis.get_dataset_info(
+            hubmap_id, instance="prod", token=token, overwrite=False
+        )
+        return datetime.fromtimestamp(metadata["published_timestamp"] / 1000.0)
+    except:
+        return None
+
+
+def __get_created_timestamp(hubmap_id, token=None):
+    try:
+        metadata = apis.get_dataset_info(
+            hubmap_id, instance="prod", token=token, overwrite=False
+        )
+        return datetime.fromtimestamp(metadata["created_timestamp"] / 1000.0)
+    except:
+        return None
+
+
+def __get_group_name(hubmap_id, token=None):
+    metadata = apis.get_dataset_info(hubmap_id, instance="prod", token=token)
+    return metadata["group_name"]
+
+
+def __get_data_type(hubmap_id, token=None):
+    metadata = apis.get_dataset_info(hubmap_id, instance="prod", token=token)
+    return metadata["data_types"]
+
+
+def __get_dataset_type(hubmap_id, token=None):
+    return apis.get_dataset_type(hubmap_id, instance="prod", token=token)
+
+
+def daily(token: str) -> pd.DataFrame:
+    """
+    Creates daily report and returns a helpful dataframe
+    """
+
+    ncores = 16
+    pandarallel.initialize(progress_bar=True, nb_workers=ncores)
+
+    utilities.pprint("Getting list of assay types")
+    assay_types = apis.get_assay_types(token=token)
+
+    print("Getting list of dataset IDs")
+    datasets = []
+    for assay_type in tqdm(assay_types):
+        hubmap_ids = apis.get_ids(assay_type, token=token)
+        datasets.extend(hubmap_ids)
+
+    df = pd.DataFrame(datasets)
+
+    utilities.pprint("Getting group name")
```

Comment:
Rather than making all these requests in a row (lines 72-92), I suggest consolidating the request logic into a single request e.g. 
- by putting the `apis.get_dataset_info` in its own function and passing its value to the double-underscore functions, or
- by employing a class-based approach and grabbing the dataset info in a cached property which can then be used by individual methods (`__is_protected`, `__get_published_timestamp`, etc.)

Reply:
(I realize I don't know enough about parallel_apply to fully account for that in a changed approach.)

Reply:
@gesinaphillips it is a great idea and I have plans for it in the future for today this suffices.